### PR TITLE
Add Closed Loop Control

### PR DESCRIPTION
**Pull request by rcya1**
Currently only has positional PID and description of feedforward. Will add more later

### CODE REVIEW

**Comment by rcya1** [comment]
@EthanMuchnik @matth101 Please review

File: docs/frc/3-Closed_Loop_Control/3b-SPARK_MAX_PID.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    private CANEncoder primaryEncoder;
+
+    private State state = defaultState;
+    private double speed;
+
+    public Elevator() {
+        primaryMotor = new CANSparkMax(ELEVATOR_PRIMARY_MOTOR_ID, MotorType.kBrushless);
+        primaryMotor.restoreFactoryDefaults();
+        primaryMotor.configSelectedFeedbackSensor(FeedbackDevice.CTRE_MagEncoder_Relative);
+        primaryMotor.configSelectedFeedbackCoefficient(1 / 4096.0 / 48.0 * Math.PI * 6);
```
Comment (matth101) [inline docs/frc/3-Closed_Loop_Control/3b-SPARK_MAX_PID.md]: I believe these commands don't go with the SPARK MAX program, just for the TalonSRX.

File: docs/frc/3-Closed_Loop_Control/3c-PID_Command_And_Extras.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+Sometimes, we want to use our PID to maintain a position even after the movement has ended or to make it maintain the same position. For instance, if we have an arm or elevator that falls under gravity, we might not want to end PID when we reach the endpoint of our movement.
+
+When this happens, we can simple remove the clause in our `PID` state that ends the `PID` state if the setpoint is reached. This will allow our closed loop control to persist.
+
+However, one should be careful with this to ensure that there is still a method available to exit the PID state. With the system described here, the release of our button is a way to end PID control. There are many other methods such as having manual control interrupt it.
+
+For freezing mechanisms, or making them maintain the same position, it is pretty easy. We follow the above steps to remove the end `PID` state clause, and then when we call the `setPosition()` function from our freeze command, we simply set the PID position to the current position. A really easy way to implement the `RobotContainer` calling of the freeze command would be to use the `toggleWhenPressed()` function when assigning buttons to commands.
+
```
Comment (matth101) [inline docs/frc/3-Closed_Loop_Control/3c-PID_Command_And_Extras.md:151]: I think it'd be nice if we mentioned our 2019 arm code as an example for freezing a mechanism (https://github.com/FRC1257/2019-Robot-Command/blob/master/src/main/java/frc/robot/subsystems/cargointake/CargoArm.java)

File: docs/frc/3-Closed_Loop_Control/4-Feedforward.md [hunk 3]
```diff
@@ -1 +1,50 @@
...
+
+```java
+double maxSpeed = 5000; // in rpm
+double desiredSpeed = 4000; // in rpm
+double kFF = 1 / maxSpeed;
+
+motor.set(desiredSpeed * kFF);
+```
+
+The other type of feedforward control takes into accounts the dynamics, or movement, of the system. We could calculate this mathematically thrugh physics, or empirically through testing, but we could do something such as counteract the effects of gravity on a system. For instance, gravity will always pull an elevator down, but we could use feedforward to counteract this by putting a constant force going up. This would prevent the elevator from naturally falling down, as well as going faster downwards than upwards.
```
Comment (matth101) [inline docs/frc/3-Closed_Loop_Control/4-Feedforward.md]: I'm not sure what you were trying to say in the second sentence of this line, maybe reword it

File: docs/frc/3-Closed_Loop_Control/4-Feedforward.md [hunk 4]
```diff
@@ -1 +1,50 @@
...
+double input = controller.getLeftY();
+
+motor.set(input + feedForward);
+```
+
+## Combining with Feedback Control
+
+However, feedforward control is rarely enough to completely make a system responsive. In fact, it works best when feedforward control is used in conjunction with feedback control. For instance, if we wanted to control the velocity of our system, we could use feedforward control as the main driving force to get our system to the desired speed, and then use feedback control to correct any errors or fluctuations in that speed.
+
+Additionally, we could use feedback control to run PID on something like an elevator subsystem that will drive our position to a desired setpoint. To counteract the effects of gravity on the system, we use feedforward to make our motor apply an upward force. This would actually help linearize the model of our system (don't worry if you don't know what this means, it just means that it will make it simpler and nicer to control), as well as remove the need for separate PID gains going up and down.
```
Comment (matth101) [inline docs/frc/3-Closed_Loop_Control/4-Feedforward.md]: Did you mean to say "Additionally, we could use _feedforward_ control...."

File: docs/frc/3-Closed_Loop_Control/4a-Velocity_PID.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+1. Set all terms but `kFF` to `0`
+2. Set `kFF` to `1 / MAX_SPEED`
+3. Tune only the `kFF` value as much as possible until you get a pretty good approximation of the desired value. Using a graph is **crucial** for this step. To do so, you need to output both the target velocity and the current velocity in the same graph.
+4. Once you've gotten the best results possible with just `kFF`, gradually increase `kP` to account for small fluctuations until you get the desired results.
+
+## Velocity Control
+
+We can not only use velocity PID to obtain a desired set speed, but also for better control. With our typical methods of control, we pass the controller joystick output to the the motor directly, which will move the motor. However, this can be imprecise and the motor could not track the desired velocity perfectly. Especially for drivetrains, we may choose to use velocity control on our subsystem instead, where we convert our joystick values to a desired velocity and send that to our velocity PID. This is generally referred to as **closed loop control** compared to the typical **open loop control**.
```
Comment (matth101) [inline docs/frc/3-Closed_Loop_Control/4a-Velocity_PID.md]: Did you mean to say "this is generally referred to as **velocity control** compared to......" ?

File: docs/frc/3-Closed_Loop_Control/5-Motion_Profiling.md [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+
+This concludes the triangular profile, and now we can go onto the trapezoidal profile. For the trapezoidal profile, we know that we can safely accelerate to our maximum velocity, cruise at that velocity for some time, and then deccelerate. To calculate how long it will take to accelerate and deccelerate, we can use the kinematic equations again.
+
+$$v_max = 0 + at_a \rightarrow t_a = t_d = \frac{v_max}{a}$$
+
+We also need to calculate the distance that we travelled during this time, which can be done with the following equation:
+
+$$v_max^2 = 0^2 + 2ax \rightarrow x = \frac{v_max^2}{2a}$$
+
+We then know that we will need to cruise at our max velocity for the remainder of the distance, which means that we will need to cruise for the time
```
Comment (matth101) [inline docs/frc/3-Closed_Loop_Control/5-Motion_Profiling.md]: There's a small issue here with the LaTeX, where only the letter m of "max" is written in subscript

**Review by matth101 — COMMENTED** [review]
The rest looks good to me, great work!

File: docs/frc/3-Closed_Loop_Control/4-Feedforward.md (hunk 4 above)
Comment (rcya1) [reply docs/frc/3-Closed_Loop_Control/4-Feedforward.md]: Actually, no b/c PID = feedback. I meant to combine feedback and feedforward control but I'll make it more clear

File: docs/frc/3-Closed_Loop_Control/4a-Velocity_PID.md (hunk 5 above)
Comment (rcya1) [reply docs/frc/3-Closed_Loop_Control/4a-Velocity_PID.md]: Not really, maybe I could put it as closed loop velocity control.
